Upload code coverage of unit tests to codecov

akshitgrover opened the pull request:
Code coverage is not being tracked for unittests, Tracking code coverage is important as it give insights on where to test and what to test

This PR updates unit tests workflow to create coverage reports and upload the same to codecov using codecov github action

Fixes #6289

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

akshitgrover commented:
@alpeb 
Please review if this solves the code coverage tracking issue #6289

alpeb commented on .github/workflows/unit_tests.yml:
```diff
@@ -23,7 +23,12 @@ jobs:
-        gotestsum -- -cover -race -v -mod=readonly ./...
+        gotestsum -- -cover -coverprofile=coverage.out -race -v -mod=readonly ./...
+    - name: Upload coverage to Codecov
+      uses: codecov/codecov-action@v1
```
It's recommended to refer directly to the sha vs the tag (also please change in the other reference below):
```suggestion
      uses: codecov/codecov-action@29386c70ef20e286228c72b668a06fd0e8399192
```

alpeb commented on .github/workflows/unit_tests.yml:
```diff
@@ -23,7 +23,12 @@ jobs:
+      uses: codecov/codecov-action@v1
+      with:
+        files: ./coverage.out
+        flags: unittests,golang
```
The report shows:
```

Flag | Coverage Δ | &nbsp;
-- | -- | --
golang | 42.44% &lt;0.00%&gt; (?) | &nbsp;
javascript | 66.69% &lt;0.00%&gt; (?) | &nbsp;
unittests | 44.37% &lt;0.00%&gt; (?)
```
These are all unit tests, so I think the `unittests` flag is redundant.

alpeb reviewed (commented):
Thanks @akshitgrover
I've triggered the tests, so you can check the results in the PR :wink:

akshitgrover commented:
@alpeb 
I have updated the workflow to use SHA

I thought of adding the 'unittests' flag as it will give the overall coverage of the project whilst 'golang' and 'javascript' projects will point to coverage of .go and .js files respectively.
Removed the 'unittests' flag for now

alpeb reviewed (approved):
LGTM :+1: :slightly_smiling_face:

akshitgrover commented:
Thank you @alpeb 
Who else should I ask for the review?
As it requires 2 people to review before I can merge

kleimkuhler reviewed (commented):
@akshitgrover thanks for adding this! Could you actually add back the `unittests` flag to the golang and javascript tests? It'll be nice to keep that as a general flag so that `unittests` represents the total coverage of all unit tests.

akshitgrover commented:
@kleimkuhler 
I have added back 'unittests' flag

kleimkuhler reviewed (approved):
Thanks @akshitgrover 🙂
